lib/Dialect/Secret/Transforms/DistributeGeneric.cpp

```diff
@@ -163,11 +169,28 @@ struct SplitGeneric : public OpRewritePattern<GenericOp> {
       }
       LoopLikeOpInterface clonedLoop =
           dyn_cast<LoopLikeOpInterface>(rewriter.clone(opToDistribute, mp));
-      for (auto [operand, blockArg] :
-           llvm::zip(clonedLoop.getInits(), clonedLoop.getRegionIterArgs())) {
-        if (isa<SecretType>(operand.getType()))
-          blockArg.setType(operand.getType());
+
+      DenseMap<Value, Value> newInitsToOperands;
+      for (auto [operand, blockArg] : llvm::zip(
+               clonedLoop.getInitsMutable(), clonedLoop.getRegionIterArgs())) {
+        auto operandNum = operand.getOperandNumber();
+        if (isa<SecretType>(operand.get().getType())) {
+          blockArg.setType(operand.get().getType());
+        } else if (solver
+                       ->lookupState<SecretnessLattice>(
+                           loop.getYieldedValues()[operandNum])
+                       ->getValue()
+                       .getSecretness() &&
+                   !isa<SecretType>(operand.get().getType())) {
+          rewriter.setInsertionPoint(clonedLoop);
+          auto newInit = rewriter.create<secret::ConcealOp>(genericOp.getLoc(),
```

Comment:
I'm a bit surprised to see secret.conceal here. Can't we just have the distributed generic pull in the non-secret value from the enclosing scope?

IMO `conceal` is only useful for a client who is encrypting a value in a scheme-agnostic manner. Not for a trivial encryption or an plaintext-ciphertext absorption.

Reply:
> Can't we just have the distributed generic pull in the non-secret value from the enclosing scope?

Hmm I don't really get it - this is the initial iter_arg (e.g. the initial value of a running sum). If the inner sum (taking the simple_sum as an example) is a ciphertext value, then how do we make the type of the initial arg match with the yielded (secret) type?

The distributed generic is inside the for loop - it's yielded value is a secret.